jetty-server/src/main/java/org/eclipse/jetty/server/Request.java

```diff
@@ -410,12 +416,55 @@ public PushBuilder newPushBuilder()
             id = getRequestedSessionId();
         }
 
+        Map<String,String> cookies = new HashMap<>();
+        Cookie[] existingCookies = getCookies();
+        if (existingCookies != null)
+        {        
+            for (Cookie c: getCookies())
+            {
+                cookies.put(c.getName(), c.getValue());
+            }
+        }
+
+        //Any Set-Cookies that were set on the response must be set as Cookies on the
+        //PushBuilder, unless the max-age of the cookie is <= 0
+        HttpFields responseFields = getResponse().getHttpFields();
+        for (HttpField field : responseFields)
+        {
+            HttpHeader header = field.getHeader();
+            if (header == null)
+                continue;
+            switch (header)
+            {
+                case SET_COOKIE:
+                {
+                    HttpCookie cookie = ((SetCookieHttpField)field).getHttpCookie();
+                    if (cookie.getMaxAge() > 0)
+                        cookies.put(cookie.getName(), cookie.getValue());
+                    else
+                        cookies.remove(cookie.getName());
+                    continue;
+                }
+                default:
+                    continue;
+            }
+        }
+        
+        if (!cookies.isEmpty())
+        {
+            StringBuilder buff = new StringBuilder();
+            for (Map.Entry<String,String> entry : cookies.entrySet())
+            {
+                if (buff.length() > 0)
+                    buff.append("; ");
+                buff.append(entry.getKey()).append('=').append(entry.getValue());
+            }
+            fields.add(new HttpField(HttpHeader.COOKIE, buff.toString()));
+        }
+        
```

Comment:
I'm not sure handling these things at `PushBuilder` creation is the correct moment to do it.
Is the TCK testing for this?

I would rather add cookies, authorization headers, etc. just before the call to `push()`, to integrate also what the application is providing.

For example, if the application calls `pushBuilder.addHeader("cookie", "a=b")` is it added to the ones computed above, or it replaces them, etc.?

Reply:
But the tck test checks that the Cookie header is set on the PushBuilder before push() is called. See https://github.com/eclipse-ee4j/jakartaee-tck/blob/8.0.2/src/com/sun/ts/tests/servlet/spec/serverpush/TestServlet4.java#L48

Reply:
> But the tck test checks that the Cookie header is set on the PushBuilder before push() is called.

That is so wrong.

Then the application calls `addHeader("Cookie", ...)` completely invalidating the fact that the `Cookie` header must be only one, so the server will complain that it found 2 Cookie headers.
Also, the application may mess with all other headers that you are trying to cope with here.

I think all this should be done just before `push()`.

I would challenge the TCK test.

Reply:
I think servers are forgiving of multiple Cookie headers.  I know we are.
I think it is reasonable that a pushed request has the same cookies as the source request and some of them may be beyond the knowledge of the app doing the push.
The builder is only setting the initial headers and the app is free to change them as it likes, so I think starting with the known cookies is reasonable.

Reply:
All right then.